tensorflow/stream_executor/cuda/cuda_dnn.cc

```diff
@@ -3259,12 +3348,50 @@ port::Status CudnnSupport::DoBatchNormalizationForwardImpl(
       batch_var_opaque = nullptr;
     }
 
-    RETURN_IF_CUDNN_ERROR(cudnnBatchNormalizationForwardTraining(
-        cudnn.handle(), mode, &one, &zero, x_descriptor.handle(), x.opaque(),
-        x_descriptor.handle(), y->opaque(), scale_offset_descriptor.handle(),
-        scale.opaque(), offset.opaque(), 1.0, batch_mean_opaque,
-        batch_var_opaque, epsilon, saved_mean->opaque(),
-        saved_inv_var->opaque()));
+    if (reserve_space_allocator != nullptr) {
+#if CUDNN_VERSION >= 7402
+      RETURN_IF_CUDNN_ERROR(cudnnBatchNormalizationForwardTrainingEx(
+          /*handle=*/cudnn.handle(),
+          /*mode=*/mode,
+          /*bnOps=*/bn_ops,
+          /*alpha=*/&one,
+          /*beta=*/&zero,
+          /*xDesc=*/x_descriptor.handle(),
+          /*xData=*/x.opaque(),
+          /*zDesc=*/NULL,
+          /*zData=*/NULL,
+          /*yDesc=*/x_descriptor.handle(),
+          /*yData=*/y->opaque(),
+          /*bnScaleBiasMeanVarDesc=*/scale_offset_descriptor.handle(),
+          /*bnScaleData=*/scale.opaque(),
+          /*bnBiasData=*/offset.opaque(),
+          /*exponentialAverageFactor=*/1.0,
+          /*resultRunningMeanData=*/batch_mean_opaque,
+          /*resultRunningVarianceData=*/batch_var_opaque,
+          /*epsilon=*/epsilon,
+          /*saveMean=*/saved_mean->opaque(),
+          /*saveInvVariance=*/saved_inv_var->opaque(),
+          /*cudnnActivationDescriptor_t=*/NULL,
+          /*workspace=*/workspace.opaque(),
+          /*workSpaceSizeInBytes=*/workspace.size(),
+          /*reserveSpace=*/reserve_space.opaque(),
+          /*reserveSpaceSizeInBytes=*/reserve_space.size()));
+#else
+      RETURN_IF_CUDNN_ERROR(cudnnBatchNormalizationForwardTraining(
+          cudnn.handle(), mode, &one, &zero, x_descriptor.handle(), x.opaque(),
+          x_descriptor.handle(), y->opaque(), scale_offset_descriptor.handle(),
+          scale.opaque(), offset.opaque(), 1.0, batch_mean_opaque,
+          batch_var_opaque, epsilon, saved_mean->opaque(),
+          saved_inv_var->opaque()));
+#endif
+    } else {
+      RETURN_IF_CUDNN_ERROR(cudnnBatchNormalizationForwardTraining(
```

Comment:
This seems to be identical to the above `cudnnBatchNormalizationForwardTraining` call in the `#else` branch, merge them?

Maybe something like:
```
    bool called = false;
#if CUDNN_VERSION >= 7402
    if (reserve_space_allocator != nullptr) {
      called = true;
      ...
    }
#endif 
    if (!called) {
      ...
    }
```